### ISSUE TITLE
Security: PBEM/PBF passwords are persisted without encryption

### ISSUE THREAD

**Issue by ssoloff** [triplea-game/triplea#1884]
I just discovered that the PBEM password (i.e. the `GenericEmailSender#m_password` field) is persisted unencrypted to the `~/triplea/local.cache` file.

It looks like there is code to prevent saving the PBEM password to a save game file.  There's even [this comment](https://github.com/triplea-game/triplea/blob/ccb6a56c287e9784bab61a0797cb449d5378d186/src/main/java/games/strategy/engine/framework/startup/ui/PBEMSetupPanel.java#L337-L338) that indicates the author knowingly saved the password to the cache unencrypted:

```java
// create a clone, delete the sensitive information in the clone, and use it in the game
// the locally cached version still has the password so the user doesn't have to enter it every time
```

Personally, I hate not knowing when an app is storing my sensitive information, much less when it's unencrypted.  I would suggest adding a warning to the UI that the user's password will be stored unencrypted and possibly giving them the ability to opt-out of persisting their password completely.  Of course, a solution that uses encryption could also be used, but then we have to discuss key management. :smile:

**Comment by panther2** [comment]
Weird, indeed. I hope we find a solution quickly. I don't like to see my forum account information being stored unencryptedly, either.
Thank you for identifying this...

**Comment by ssoloff** [comment]
Thanks for mentioning "forum," @panther2.  I confirmed that all PBF providers are also affected by this issue.  In this case, the password is coming from `AbstractForumPoster#m_password`.  It basically has the same implementation as `GenericEmailSender` with respect to ensuring the password is not stored in save game files but storing it unencrypted to `local.cache`.  Therefore, PBF should also be addressed in the fix.

**Comment by ron-murhammer** [comment]
Not great and definitely would be nice to encrypt it. Though I'm glad its just on the local machine as that is much less of a concern than say a saved game.

**Comment by RoiEXLab** [comment]
I'd suggest handling this problem similar to web browsers like firefox or chrome.
They simply store passwords encrypted, along with an unencrypted key unless using a master password.
A master password to store a single password (or 2) is pretty useless so we should just encrypt the password with a random key we store elsewhere on the system

**Comment by ssoloff** [comment]
I just discovered another persistent copy of the user credentials when using `TripleAForumPoster`.  This class has two non-transient fields for `username` and `password`, which are being saved without encryption to `local.cache` (however, the password _is_ scrubbed from save game files).

@RoiEXLab, do you foresee any problem with inlining these fields into the code, or at least making them `transient`?  Since this feature was added after the last release, do we need to worry about compatibility?

If we can inline them, this seems to work:

```diff
diff --git a/src/main/java/games/strategy/engine/pbem/TripleAForumPoster.java b/src/main/java/games/strategy/engine/pbem/TripleAForumPoster.java
index cbf8f43..d43b9dd 100644
--- a/src/main/java/games/strategy/engine/pbem/TripleAForumPoster.java
+++ b/src/main/java/games/strategy/engine/pbem/TripleAForumPoster.java
@@ -33,17 +33,13 @@ public class TripleAForumPoster extends AbstractForumPoster {
 
   private static final String tripleAForumURL = "https://forums.triplea-game.org";
 
-  private NameValuePair username;
-  private NameValuePair password;
-
-
   @Override
   public boolean postTurnSummary(final String summary, final String title) {
-    username = new BasicNameValuePair("username", getUsername());
-    password = new BasicNameValuePair("password", getPassword());
+    final NameValuePair username = new BasicNameValuePair("username", getUsername());
+    final NameValuePair password = new BasicNameValuePair("password", getPassword());
     try (CloseableHttpClient client = HttpClients.custom().disableCookieManagement().build()) {
-      final int userId = getUserId(client);
-      final String token = getToken(client, userId);
+      final int userId = getUserId(client, username, password);
+      final String token = getToken(client, userId, password);
       try {
         post(client, token, "### " + title + "\n" + summary);
         m_turnSummaryRef = "Sucessfully posted!";
@@ -95,7 +91,8 @@ public class TripleAForumPoster extends AbstractForumPoster {
     client.execute(httpDelete);
   }
 
-  private int getUserId(final CloseableHttpClient client) throws Exception {
+  private static int getUserId(final CloseableHttpClient client, final NameValuePair username,
+      final NameValuePair password) throws Exception {
     final JSONObject jsonObject = login(client, Arrays.asList(username, password));
     checkUser(jsonObject);
     return jsonObject.getInt("uid");
@@ -124,7 +121,8 @@ public class TripleAForumPoster extends AbstractForumPoster {
     }
   }
 
-  private String getToken(final CloseableHttpClient client, final int userId) throws Exception {
+  private static String getToken(final CloseableHttpClient client, final int userId, final NameValuePair password)
+      throws Exception {
     final HttpPost post = new HttpPost(tripleAForumURL + "/api/v1/users/" + userId + "/tokens");
     post.setEntity(new UrlEncodedFormEntity(Collections.singletonList(password), StandardCharsets.UTF_8));
     HttpProxy.addProxy(post);
```

**Comment by ssoloff** [comment]
@RoiEXLab Bumping previous comment...  This is the last problem that needs to be fixed before we can close this issue.

**Comment by RoiEXLab** [comment]
@ssoloff I don't have any problems with that, however I want to point out a few things...
Currently the forum Posters are created via reflection.
(I believe this is in order to restore them at some point (see `PBEMSetupPanel#findCachedOrCreateNew`)).
If we replace the current system, there might not even be a need to remove this field at all...

**Comment by ssoloff** [comment]
@RoiEXLab 

> If we replace the current system, there might not even be a need to remove this field at all...

Possibly.  Right now the problem isn't the reflection so much as it's that they are `Serializable` and that is how they are persisted to `local.cache`.  Currently, since `TripleAForumPoster#username` and `TripleAForumPoster#password` are not transient, they get persisted, and that becomes another attack surface we have to worry about.

I agree that the current `IBean` infrastructure feels a little clunky and can be improved.  However, to immediately address the security issue, I think it will be an order of magnitude easier to do a 2-12 line change in one class than to redesign how the entire `PBEMSetupPanel` works. :smile:

**Comment by ssoloff** [comment]
@ron-murhammer @DanVanAtta 

I believe all the concerns raised in this issue have been addressed by #1929, #1955, #2000, and #2003.  In summary:

* By default, we do not store the user's PBEM/PBF credentials.  The user must explicitly opt-in by selecting the **Remember me** option in the PBEM/PBF panel.
* If the user chooses to have us remember their credentials, they will be stored encrypted in the local cache using a user-specific randomly-generated master password.  The master password is stored unencrypted in a location that is only readable by the user.
* User credentials are **never** stored in a save game, encrypted or otherwise.
